sdk-tools/src/sign.rs

```diff
@@ -0,0 +1,272 @@
+        }
+
+        if self.ignore_init_section_error {
+            cmd.arg("-ignore-init-sec-error");
+        }
+
+        if self.resign {
+            cmd.arg("-resign");
+        }
+
+        cmd
+    }
+
+    /// Combine an unsigned enclave and signature into the output enclave, after
+    /// checking the signature.
+    pub fn catsig(
+        &mut self,
+        unsigned_enclave: &Path,
+        config_path: &Path,
+        public_key_pem: &Path,
+        gendata_output: &Path,
+        signature: &Path,
+        output_enclave: &Path,
+    ) -> Command {
+        let mut cmd = Command::new(self.sgx_sign_path.clone());
+        cmd.arg("catsig")
+            .arg("-enclave")
+            .arg(unsigned_enclave)
+            .arg("-config")
+            .arg(config_path)
+            .arg("-key")
+            .arg(public_key_pem)
+            .arg("-unsigned")
+            .arg(gendata_output)
+            .arg("-sig")
+            .arg(signature)
+            .arg("-out")
+            .arg(output_enclave);
+
+        if self.ignore_relocation_error {
+            cmd.arg("-ignore-rel-error");
+        }
+
+        if self.ignore_init_section_error {
+            cmd.arg("-ignore-init-sec-error");
+        }
+
+        cmd
+    }
+
+    /// Examine a signed enclave file and dump the data
+    pub fn dump(
+        &mut self,
+        signed_enclave: &Path,
+        css_file_path: &Path,
+        dump_file_path: &Path,
+    ) -> Command {
+        let mut cmd = Command::new(self.sgx_sign_path.clone());
+        cmd.arg("dump")
+            .arg("-enclave")
+            .arg(signed_enclave)
+            .arg("-dumpfile")
+            .arg(dump_file_path)
+            .arg("-cssfile")
+            .arg(css_file_path);
+
+        if self.ignore_relocation_error {
+            cmd.arg("-ignore-rel-error");
+        }
+
+        if self.ignore_init_section_error {
+            cmd.arg("-ignore-init-sec-error");
+        }
+
+        if self.resign {
+            cmd.arg("-resign");
+        }
+
+        cmd
+    }
+}
+
+/// Construct a new SgxSign utility around the given executable path
+impl From<PathBuf> for SgxSign {
+    fn from(sgx_sign_path: PathBuf) -> Self {
+        Self {
+            sgx_sign_path,
+            ignore_relocation_error: false,
+            ignore_init_section_error: false,
+            resign: false,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_generate_sign_command() {
+        let cmd = format!(
+            "{:?}",
+            SgxSign::default().sign(
+                Path::new("unsigned_enclave"),
+                Path::new("config_path"),
+                Path::new("private_key"),
+                Path::new("output_enclave"),
+            )
+        );
+        let expected_cmd = "\"/opt/intel/sgxsdk/bin/x64/sgx_sign\" \"sign\" \"-enclave\" \"unsigned_enclave\" \"-config\" \"config_path\" \"-key\" \"private_key\" \"-out\" \"output_enclave\"";
```

Comment:
:exploding_head: This is an awesome way to test Command
I had written some tests for Command and I ended up doing some nasty trait logic so I could mock out Command, it was ugly and intrusive.
This idea of having a method (preferably private) that returns the command and comparing the string, then either not testing the final execution or leaving it to something like [assert_cmd](https://docs.rs/assert_cmd/latest/assert_cmd/) is so much cleaner.

Reply:
💯